### PR TITLE
[dynamo] support torchbind object input

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124978

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: test/cpp/jit/test_custom_class_registrations.cpp [hunk 1]
```diff
@@ -199,6 +203,11 @@ struct TensorQueue : torch::CustomClassHolder {
     return raw_queue;
   }
 
+  std::tuple<std::tuple<std::string, std::vector<at::Tensor>>> __obj_flatten__() {
+    // for demostration purpose, we also return a size integer.
```
Comment (ydwu4) [inline test/cpp/jit/test_custom_class_registrations.cpp:207]: change comment

File: torch/_dynamo/codegen.py [hunk 2]
```diff
@@ -143,6 +144,7 @@ def __call__(self, value, allow_cache=True):
                 SymNodeVariable,
                 UnspecializedPythonVariable,
                 NumpyNdarrayVariable,
+                TorchScriptObjectVariable,
```
Comment (ydwu4) [inline torch/_dynamo/codegen.py:147]: test or remote it

File: torch/_dynamo/output_graph.py [hunk 3]
```diff
@@ -1423,7 +1423,11 @@ def remove_unused(node):
             self.remove_node(node)
             self.real_value_cache.pop(node, None)
 
-        used_symbols = set()
```
Comment (ydwu4) [inline torch/_dynamo/output_graph.py:1426]: test this

File: torch/_dynamo/variables/builder.py [hunk 4]
```diff
@@ -888,6 +890,38 @@ def build_key_value(i, k, v):
                     user_cls_source=AttrSource(self.source, "__class__"),
                 ),
             )
+        elif TorchScriptObjectVariable.is_matching_cls(type(value)):
```
Comment (ydwu4) [inline torch/_dynamo/variables/builder.py:893]: Test what happens if they're closure or global.

File: torch/_dynamo/variables/script_object.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    def __init__(self, proxy, value, source, **kwargs):
+        super().__init__(value, **kwargs)
+        self.proxy = proxy
+        self.proxy.node.meta["example_value"] = value
+        self.source = source
+
+    def as_proxy(self):
+        return self.proxy
+
+    def var_getattr(self, tx, name: str) -> VariableTracker:
```
Comment (ydwu4) [inline torch/_dynamo/variables/script_object.py:29]: test what happens if getattr, setattr to TorchScriptObjectVariable, call method that doesn't exist and ban the setattrs (check the previous behavior of torch.ScriptObject).

File: torch/_export/passes/lift_constants_pass.py [hunk 6]
```diff
@@ -244,10 +244,8 @@ def lift_constants_pass(
 def rewrite_script_object_meta(
     gm: torch.fx.GraphModule,
 ) -> Dict[str, Union[torch.Tensor, FakeScriptObject],]:
-    """When tracing, we produce a graph with FakeScriptObject in the
-    meta["val"].
-
-    For now, we rewrie meta["val"] to be a placeholder CustomObjArgument
+    """When tracing, we produce a graph with an FakeScriptObject in the
```
Comment (ydwu4) [inline torch/_export/passes/lift_constants_pass.py:247]: remote the random change

File: torch/_library/fake_class_registry.py [hunk 7]
```diff
@@ -63,8 +63,35 @@ def _check_registered(self, full_qualname: str) -> None:
...
+
+        if not len(tp) == 2 or not isinstance(tp[0], str):
+            raise RuntimeError(
+                "Expect element of flat x to be a tuple of two elements with first element being a string"
+            )
+
+
 def to_fake_obj(fake_mode, x: torch.ScriptObject) -> FakeScriptObject:
-    fake_x = _fake_obj_from_real(fake_mode, x)
+    import torch.utils._pytree as pytree
```
Comment (ydwu4) [inline torch/_library/fake_class_registry.py:82]: add tests.

File: torch/_library/fake_class_registry.py [hunk 8]
```diff
@@ -87,7 +114,8 @@ def wrapped(self_, *args, **kwargs):
                 _call_torchbind(name).__get__(fake_x_wrapped),
             )
         else:
-            log.warning("fake object of %s doesn't implement method %s.", x, name)
+            # log.warning("fake object of %s doesn't implement method %s.", x, name)
```
Comment (ydwu4) [inline torch/_library/fake_class_registry.py:117]: remove the change

File: torch/_functorch/aot_autograd.py [hunk 9]
```diff
@@ -500,6 +500,8 @@ def convert(idx, x):
                         return shape_env.create_symintnode(
                             shape_env.create_symbol(x, source), hint=x, source=source
                         )
+                if isinstance(x, torch.ScriptObject):
+                    return torch._library.fake_class_registry.to_fake_obj(fake_mode, x)
```
Comment (ydwu4) [inline torch/_functorch/aot_autograd.py:504]: test this.

File: torch/_dynamo/variables/builder.py [hunk 10]
```diff
@@ -1721,6 +1755,11 @@ def _clone_input(value):
     ]:
         set_example_value(proxy.node, example_value)
         return ConstantVariable.create(example_value, **options)
+    elif (
+        isinstance(example_value, (int, float, bool))
+        and proxy.node.target is torch._higher_order_ops.torchbind.call_torchbind
```
Comment (ydwu4) [inline torch/_dynamo/variables/builder.py:1760]: unit test for these

File: torch/_dynamo/output_graph.py [hunk 11]
```diff
@@ -1441,10 +1445,22 @@ def remove_unused(node):
                     arg = node.meta["grapharg"]
                     if isinstance(arg, BackwardStateGraphArg):
                         continue
+                    if isinstance(node.meta["grapharg"].example, torch.ScriptObject):
```
Comment (ydwu4) [inline torch/_dynamo/output_graph.py:1448]: animesh: check whether methods are called of script object and decide whether to delete it or not.

File: torch/_dynamo/variables/script_object.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+            call_torchbind,
+            source=AttrSource(self.source, name),
+            script_obj_var=self,
+            method_name=name,
+        )
+
+    @_raise_hard_error_if_graph_break(
+        "Dynamo cannot safely trace script object due to graph break."
+    )
+    def call_method(self, tx, name, args, kwargs):
```
Comment (ydwu4) [inline torch/_dynamo/variables/script_object.py:75]: animesh: forbid anything that directly call call_method e.g. __get_attr__

File: torch/_dynamo/output_graph.py (hunk 11 above)
Comment (ydwu4) [reply torch/_dynamo/output_graph.py:1448]: It turns out the current implementation might be correct: used_symbols is not used to prune the script object (it's pruned in 1439) but used to prune SymInt intputs. In that case , we should add the symints in the contents of torch.ScriptObject to used_symbols to be safe.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**